### ISSUE TITLE
How to take backup to AWS S3 persistence store 

### ISSUE THREAD

**Issue by abhisamant7** [hyfen-nl/PIVT#42]
Hi Raft,

I have been gone though  PIVT project which is great help who wants to setup production HLF. you have done great job. thanks for this project.

I have query regarding the backup section in this project. I want to backup on AWS S3 instead of Azure Blob Storage, however in your project it mentioned only Azure Blob Storage so How can I do it on AWS so I can take backup and Restore it .

Thanks 
Abhijeet

**Comment by raftAtGit** [comment]
Thanks for the nice words :)

See the bottom of [backup-workflow.yaml](https://github.com/APGGroeiFabriek/PIVT/blob/master/fabric-kube/backup-flow/templates/backup-workflow.yaml#L218) please. You can easily extend it similarly for S3 using [AWS CLI image](https://hub.docker.com/r/amazon/aws-cli).

Also please make a pull request after implementing this ;)

**Comment by abhisamant7** [comment]
Oh means instead of use microsoft/azure-cli:latest I need to use amazon/aws-cli?

**Comment by raftAtGit** [comment]
yes, and implement the command using AWS CLI instead of Azure CLI

**Comment by abhisamant7** [comment]
I have one query also, like backup can only be done on simple network or we can implement on scaled-raft-tls? 
As on given example on backup section showed for simple-persistence one. If yes how can we do with scaled-raft-tls

Thanks

**Comment by raftAtGit** [comment]
you can take backup of any kind of network, as long as `persistence` is enabled. no specific action required for the type of network.

just beware, if that is a Kafka network, taking orderer backups is not enough. Kafka cluster should also be backed up, which is out of scope of this repository.

**Comment by abhisamant7** [comment]
Hi Raft,

Can you help me to change command which is mention as below to AWS S3 bucket. 
`helm template -f samples/simple-persistent/crypto-config.yaml --set backup.target.azureBlobStorage.accountName=<your account name> --set backup.target.azureBlobStorage.accessKey=<your access key> backup-flow/ | argo submit  -  --watch
`
I never tried both azure Blob and aws s3 so if you can help me to convert it to aws command instead of azure it would be great for me.

Thanks

**Comment by raftAtGit** [comment]
it's not just changing a single command. you need to basically:
1. add implementation for uploading backup stuff to AWS using AWS CLI Docker image, very similar to [Azure one](https://github.com/hyfen-nl/PIVT/blob/master/fabric-kube/backup-flow/templates/backup-workflow.yaml#L218)
2. capture required parameters, (possibly including access_key_id and secret_access_key)
3. pass that parameters to backup flow
4. run and test the flow

**Comment by raftAtGit** [comment]
@abhisamant7 try [this branch](https://github.com/hyfen-nl/PIVT/tree/feature/42-AWS-backup) please.

you basically need to set `backup.target.type` to `awsS3` and also provide bucket name, region, access key id and secret access key.

```
backup:
  target: 
    type: awsS3
    awsS3: 
      bucketName: 
      region:
      accessKeyId: 
      secretAccessKey: 
```

**Comment by abhisamant7** [comment]
Hi Raft,
Thanks for your effort to write script for AWS but 
I have pulled your changes  as mentioned and used below command to Start backup procedure
`helm template -f samples/scaled-raft-tls/crypto-config.yaml --set backup.target.awsS3.bucketName=backupretoredfarm --set backup.target.awsS3.accessKeyId=<AWSaccesskeyid> --set backup.target.awsS3.secretAccessKey=<AWSsecretAccessKey> backup-flow/ | argo submit  -  --watch`

 however when I am trying to Start backup procedure  it give me error as below which I attached 
![Screenshot from 2020-10-19 01-17-00](https://user-images.githubusercontent.com/21276072/96378400-49c93380-11a9-11eb-976b-d67e279fa0f1.png)


Please help me how can I resolve the issue

**Comment by raftAtGit** [comment]
that error looks like you didnt configure and artifact repository for Argo. easiest is to use Minio.

https://argoproj.github.io/argo/configure-artifact-repository/

**Comment by abhisamant7** [comment]
Ok So configure an artifact repository for Argo which file I needs to add it or I have to create new configuration?

**Comment by raftAtGit** [comment]
see the [configure-the-default-artifact-repository](https://argoproj.github.io/argo/configure-artifact-repository/#configure-the-default-artifact-repository) section

**Comment by abhisamant7** [comment]
> see the [configure-the-default-artifact-repository](https://argoproj.github.io/argo/configure-artifact-repository/#configure-the-default-artifact-repository) section

Hi Raft,

I gone though [this](configure-the-default-artifact-repository) however I am not sure how to edit configmap workflow-controller-configmap and update new changes which is added in configmap of argo. If possible could you please guide me little on this.

Thanks

**Comment by raftAtGit** [comment]
just type the command:
```
kubectl edit configmap workflow-controller-configmap -n argo
```
edit the values and save ;)

**Comment by abhisamant7** [comment]
> just type the command:
> 
> ```
> kubectl edit configmap workflow-controller-configmap -n argo
> ```
> 
> edit the values and save ;)

Hi Raft,

I have tried above command and added the stuff as below however still unable to configure it with aws s3 bucket and getting same issue as mentioned above.Now I am clueless what needs to do. Please see attached screenshot of argo configmap and please assist where I am doing wrong.
![Screenshot from 2020-10-28 18-52-16](https://user-images.githubusercontent.com/21276072/97441860-2fcce500-194f-11eb-9baf-23b581bdc1d4.png)
 

```
data:
  artifactRepository: |
    s3:
      bucket: backupretoredfarm
      keyFormat: /   
      endpoint: s3.amazonaws.com   
      insecure: true         
      useSDKCreds: true            
	  
```

**Comment by raftAtGit** [comment]
this wasnt what I meant. although you can do that, no need to setup AWS S3 artifact repository for Argo. this is not relevant to where to store Fabric backup. artifact repository is used to copy files (artifacts) from one step to another in Argo flows.

I suggest using Minio as artifact repository for Argo as it's the easiest.

I cannot provide you step by step instructions for setting it up, please just follow Argo documents and make sure you can run a sample Argo flow which uses artifacts before continuing Fabric backup.
